### ISSUE TITLE
onAuthStateChanged infinitely spawns event listeners

### ISSUE THREAD

**Issue by ganey** [firebase/firebase-js-sdk#6082]
### [REQUIRED] Describe your environment

  * Operating System version: MacOS 12.3 on M1 Pro
  * Browser version: Google Chrome Version 99.0.4844.74 (Official Build) (arm64)
  * Firebase SDK version: 9.6.8
  * Firebase Product: auth

#### Steps to reproduce:
1. clone [this repository](https://github.com/ganey/firebase-auth-listener-bug) adapted from the firebase-quickstart-js examples
2. update the `firebaseConfig` variable in index.html with your firebase details
3. Visit index.html in chrome browser & open dev tools
4. click 3 dots top right of devtools, more tools, then performance monitor
5. enable JS Event Listeners and watch the count go up infinitely

![image](https://user-images.githubusercontent.com/1401832/159018931-47902ba6-bcbd-409e-807f-1c0a4edab3f6.png)

#### Relevant Code:

Commenting out the [firebase.auth().onAuthStateChanged(function(user) {](https://github.com/firebase/quickstart-js/blob/master/auth/email-password.html#L136) prevents the listener count going up:

![image](https://user-images.githubusercontent.com/1401832/159019989-e969fa50-9e20-492e-9770-4054e25e3a7d.png)

#### Notes:

In the few minutes it's taken to create this issue, the count has already more than doubled:

![image](https://user-images.githubusercontent.com/1401832/159019788-660ef53b-9bd3-43c4-9110-692b1ab07c97.png)

**Comment by sam-gc** [comment]
Hi @ganey, thanks for filing this. This issue has come up before but it's not what it seems. The SDK polls indexedDB every 800ms in order to synchronize with other tabs (there is no widely-supported cross-tab indexedDB storage event like there is for localStorage).

IndexedDB requests require subscriptions using event listeners. In our code that happens here:
https://github.com/firebase/firebase-js-sdk/blob/69aa7b02df3b4d1f9832b7713951936b6bf32ca9/packages/auth/src/platform_browser/persistence/indexed_db.ts#L60-L73

The request itself will eventually be garbage collected, along with the listeners. I've confirmed that manually unsubscribing from both success and error listeners when either callback fires still leads to the behavior you see in the performance monitor. I suspect it's just an artifact of the garbage collector not kicking in while you're watching the graph (watching for a while on my machine, however, it does eventually fall back down after GC).

Also just for thoroughness, it's not `onAuthStateChanged` that triggers this, but merely the instantiation of the library at all. Simply calling `firebase.auth()` in the dev tools will also cause this behavior.

This is generally not something to worry about since, again, the GC will clean up. But if it's concerning, you can switch to the modular syntax and specify [`browserLocalPersistence` specifically with `initializeAuth`](https://firebase.google.com/docs/auth/web/custom-dependencies), which will ensure the indexedDB code that adds these listeners won't be executed:

```js
import {initializeAuth, browserLocalPersistence, browserPopupRedirectResolver} from 'firebase/auth';

const auth = initializeAuth(app, {
  persistence: browserLocalPersistence,
  popupRedirectResolver: browserPopupRedirectResolver,
});
```

TL;DR it's unfortunate how this appears in dev tools, but it's not actually really an issue. I'm going to close this for now, but please feel free to follow up with other questions.

**Comment by ganey** [comment]
Hey @sam-gc , thanks for clarifying how it all works and linking to the source (i started looking but got lost in the code base),  I guess I wasn't waiting long enough for GC to kick in and clean up.

I assume GC will also cleanup everything in a service worker too (i'm using an adaptation of the [Google auth service worker)](https://firebase.google.com/docs/auth/web/service-worker-sessions)?

**Comment by sam-gc** [comment]
Yep, the same should apply in service workers
